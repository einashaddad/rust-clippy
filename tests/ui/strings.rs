


#[warn(string_add)]
#[allow(string_add_assign)]
fn add_only() { // ignores assignment distinction
    let mut x = "".to_owned();

    for _ in 1..3 {
        x = x + ".";
    }

    let y = "".to_owned();
    let z = y + "...";

    assert_eq!(&x, &z);
}

#[warn(string_add_assign)]
fn add_assign_only() {
    let mut x = "".to_owned();

    for _ in 1..3 {
        x = x + ".";
    }

    let y = "".to_owned();
    let z = y + "...";

    assert_eq!(&x, &z);
}

#[warn(string_add, string_add_assign)]
fn both() {
    let mut x = "".to_owned();

    for _ in 1..3 {
        x = x + ".";
    }

    let y = "".to_owned();
    let z = y + "...";

    assert_eq!(&x, &z);
}

#[allow(dead_code, unused_variables)]
#[warn(string_lit_as_bytes)]
fn str_lit_as_bytes() {
    let bs = "hello there".as_bytes();

    // no warning, because this cannot be written as a byte string literal:
    let ubs = "â˜ƒ".as_bytes();

    let strify = stringify!(foobar).as_bytes();
}

#[warn(string_repeat_once)]
fn str_repeat() {
    let a = "this shouldn't work".repeat(1);
    let b = String::from("neither should this");
    let c = b.repeat(1);
    let d = "this works".repeat(2);
}

fn main() {
    add_only();
    add_assign_only();
    both();
    str_repeat();

    // the add is only caught for `String`
    let mut x = 1;
    ; x = x + 1;
    assert_eq!(2, x);
}
