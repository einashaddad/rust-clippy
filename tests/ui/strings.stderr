error: manual implementation of an assign operation
  --> $DIR/strings.rs:10:9
   |
10 |         x = x + ".";
   |         ^^^^^^^^^^^ help: replace it with: `x += "."`
   |
   = note: `-D assign-op-pattern` implied by `-D warnings`

error: you added something to a string. Consider using `String::push_str()` instead
  --> $DIR/strings.rs:10:13
   |
10 |         x = x + ".";
   |             ^^^^^^^
   |
   = note: `-D string-add` implied by `-D warnings`

error: you added something to a string. Consider using `String::push_str()` instead
  --> $DIR/strings.rs:14:13
   |
14 |     let z = y + "...";
   |             ^^^^^^^^^

error: you assigned the result of adding something to this string. Consider using `String::push_str()` instead
  --> $DIR/strings.rs:24:9
   |
24 |         x = x + ".";
   |         ^^^^^^^^^^^
   |
   = note: `-D string-add-assign` implied by `-D warnings`

error: manual implementation of an assign operation
  --> $DIR/strings.rs:24:9
   |
24 |         x = x + ".";
   |         ^^^^^^^^^^^ help: replace it with: `x += "."`

error: you assigned the result of adding something to this string. Consider using `String::push_str()` instead
  --> $DIR/strings.rs:38:9
   |
38 |         x = x + ".";
   |         ^^^^^^^^^^^

error: manual implementation of an assign operation
  --> $DIR/strings.rs:38:9
   |
38 |         x = x + ".";
   |         ^^^^^^^^^^^ help: replace it with: `x += "."`

error: you added something to a string. Consider using `String::push_str()` instead
  --> $DIR/strings.rs:42:13
   |
42 |     let z = y + "...";
   |             ^^^^^^^^^

error: calling `as_bytes()` on a string literal
  --> $DIR/strings.rs:50:14
   |
50 |     let bs = "hello there".as_bytes();
   |              ^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using a byte string literal instead: `b"hello there"`
   |
   = note: `-D string-lit-as-bytes` implied by `-D warnings`

error: calling `as_bytes()` on a string literal
  --> $DIR/strings.rs:55:18
   |
55 |     let strify = stringify!(foobar).as_bytes();
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using a byte string literal instead: `bstringify!(foobar)`

error: calling `repeat(1)` on a string literal
  --> $DIR/strings.rs:60:13
   |
60 |     let a = "this shouldn't work".repeat(1);
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using `.clone()` instead: `"this shouldn't work".clone()`
   |
   = note: `-D string-repeat-once` implied by `-D warnings`

error: calling `repeat(1)` on a string literal
  --> $DIR/strings.rs:62:13
   |
62 |     let c = b.repeat(1);
   |             ^^^^^^^^^^^ help: consider using `.clone()` instead: `b.clone()`

error: manual implementation of an assign operation
  --> $DIR/strings.rs:74:7
   |
74 |     ; x = x + 1;
   |       ^^^^^^^^^ help: replace it with: `x += 1`

error: aborting due to 13 previous errors

